### PR TITLE
perf(gemma4): a newcomer's prompt rides the live decode batch

### LINKED ISSUE #914: gemma4: an admission freezes every live stream for its whole prefill
## Description

With streams in flight, a new request's prefill runs as its own step before the next decode round, so every live stream freezes for the newcomer's whole prompt — and the admission loop runs up to the slot ceiling's worth of prefills per turn. Measured with a streaming request underneath a flood of one-token requests, its inter-token gap goes from ~28 ms to ~500 ms and stays there whether 16, 48 or 96 requests are queued. The serving doc records this as the line's first limit.

The step machinery no longer requires it: the decode preps can serve a row suffix by pointer offset, the local windowed prefill kernel already reads ragged batches through one plan, and the global family's decode read takes per-slot metadata. One step can carry the prompt in the row prefix and every live stream's token in the suffix.

## Proposed Plan

- Teach the decode preps and the global split decode read to serve the row suffix `[row_offset..seq_len)`; pure-decode callers pass 0.
- Add a mixed step: the admitted prompt and the live decode batch share one weight scan — one ragged local plan covers every row's windowed read, the global family keeps its two shapes (a per-prompt plan for the prompt rows, the native split entry at a pseudo-row offset for the suffix), and one sampler call covers the newcomer's first token plus every active row. Always eager; pure-decode steps keep their bucketed graph replays.
- Route admission through it whenever the decode batch is non-empty; a prompt arriving with nothing active keeps the plain path.
- Gate it on token equality with the serial path: two mixed admissions into a live batch, every lane compared whole.

## Acceptance Criteria

- The mixed gate holds: newcomer and every incumbent lane token-exact against the serial path.
- The existing gates hold: checkpoint oracles, ragged determinism, live serving checks.
- The flood measurement is re-run and recorded; the pure-decode paths show no regression.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #914

An admission used to freeze every live stream for the newcomer's whole prefill: the prompt ran as its own step before the next decode round, up to the slot ceiling's worth of prefills per turn. Under a flood of one-token requests, a live stream's inter-token gap measured ~500 ms whatever the backlog depth.

The first commit teaches the decode preps and the global split decode read to serve a row suffix `[row_offset..seq_len)` by pointer offset; pure-decode callers pass 0.

The second commit adds the mixed step and routes admission through it whenever the decode batch is non-empty. The prompt rows sit in the row prefix and every live stream's token in the suffix, sharing one weight scan: one ragged local plan covers every row's windowed read in a single launch, and the global family keeps its two shapes — the prompt rows read through a per-prompt plan, the decode suffix through the native split entry at a pseudo-row offset — so every decode row runs the same kernels and chunk plan as a pure decode step. One sampler call then covers the newcomer's first token (logits row 0) and every active row after it. Mixed steps are always eager — the prompt length varies per admission, so the shape never rides a graph — and the pure-decode steps around them keep their bucketed replays. A prompt arriving with nothing active keeps the plain path, as does `echo`.

The gate drives two mixed admissions into a live batch and demands token equality with the serial path for every lane, whole sequences compared.

### Test Env

- Single GPU (sm_89, x86_64), CUDA 12.9, Gemma 4 12B checkpoint.

### Verification

- Ten build/lint/test gates green (fmt, kernels+gemma4 builds, clippy `-D warnings` incl. the CI crate set, hd256/hd512 smokes, window-prefill reject, gemma4 lib tests).
- Mixed gate green: three lanes at budgets 50/37/44 tokens, two mixed admissions into a live batch — every lane token-exact against the serial path.
- Checkpoint oracles, ragged determinism gate and live serving checks green on the same tip.
- Flood A/B (streaming request underneath N one-token admissions; gap = the stream's inter-token latency during the flood):

| arm | flood 16: p99 / max | flood 48: p99 / max | flood 96: p99 / max | stream tokens through a 96-flood |
| --- | --- | --- | --- | --- |
| base | 203.7 / 321.0 ms | 498.8 / 499.3 ms | 498.6 / 499.4 ms | 27 |
| this PR | **30.2 / 30.5 ms** | **30.3 / 30.4 ms** | **30.3 / 30.4 ms** | 122 |

  The worst gap under any flood depth is one mixed step. Flood completion wall is unchanged (±2%), and the pure paths do not move: TPOT 27.60 vs 27.56 ms, c1 34.1 vs 34.0 tok/s; c8/c9 measured +3-5% because the fixed-set measurement includes the admission ramp, with wider run spread than the pure-decode arms. Resident is unchanged.

### CODE REVIEW

**Comment by FeathBow** [comment]
@codex review please

File: docs/models/gemma4/serving.md [hunk 1]
```diff
@@ -105,7 +105,7 @@ The consequence for callers: **greedy output is reproducible for a given workloa
...
+- **An admission rides the live decode batch instead of freezing it.** With streams in flight, a newcomer's prompt shares one eager step with the decode batch: the prompt rows sit in the row prefix, every active stream advances its token in the suffix, and one sampler call covers the newcomer's first token and every active row. Measured with a streaming request underneath a flood of one-token requests: its inter-token gap stays at about 30 ms — one mixed step — whether 16, 48 or 96 requests are queued, where the frozen-prefill scheduling this replaced measured about 500 ms at the same depths. Admission attempts per turn stay bounded by the slot ceiling; a prompt arriving with nothing active still prefills alone.
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply docs/models/gemma4/serving.md:108]: Sure, done
